### ISSUE TITLE
review feedback for ELIP 100 "Hardware wallet extensions for PSET"

### ISSUE THREAD

**Issue by philippem** [ElementsProject/ELIPs#5]

**Comment by jgriffiths** [comment]
@philippem Asked me to review this ELIP, I will add my comments as individual comments here.

**Comment by jgriffiths** [comment]
> The fields added for PSET are only allowed when the PSBT version is 2. 

Redundant, there is no PSBT version for PSET, and PSET only supports v2.

**Comment by jgriffiths** [comment]
> it is stored as 3 bytes starting with 0xfd. For example, identifier of PSBT_ELEMENTS_HWW_GLOBAL_ASSET_METADATA is a sequence of 9 bytes 0xfc0470736574fd0100. 

This sequence of bytes does not start with 0xfd.

**Comment by jgriffiths** [comment]
> Subtype of the fields belonging to the hardware wallet extensions always start with 0x0100 base to avoid potential collisions when new fields will be added to the mainstream PSET standard

PSBT/PSET already has a mechanism to avoid collisions, using the proprietary field prefix. This spec should use another field prefix than "pset" and number its fields from zero.

1. Extending the existing numbers would have broken elements prior to my fix e8e2bd1e61110f74ce7abaf2f4dce6118ee99a3a, whereas an unknown field prefix would have been no problem. I remain unsure as to whether elements will correctly roundtrip (i.e. emit after parsing) unknown field numbers under the "pset" extension. Can we expect this not to break other implementations when the original spec and reference implementation owner got it wrong?
2. All implementations can verify that they handle non "pset" prefixed data with a single test case
3. 0x100 is entirely arbitrary and was not documented up-front in the PSET spec as a range for a given purpose
4. Having large gaps between the field numbers prevents bitfield driven analysis such as wally's implementation, which uses a bitfield to detect spec violations when parsing with only a couple of machine instructions (compared to pages of code doing manual checks in e.g. elements). This is valuable because the spec's required/exclusive/versioned field requirements can be directly translated to bitflags, which found a lot of errors both in the spec and the reference implementation. It also makes the implementation suitable for very low powered devices so I'd strongly prefer not to lose this ability and the benefits it brings.

**Comment by jgriffiths** [comment]
> `<compact size uint contractLen><contract><32-byte prevoutTxid><32-bit little endian uint prevoutIndex>`

As a general rule, prefer to have fixed size data before variable length data, it makes decoding easier since otherwise all fields need extra math to adjust their offset at runtime.

**Comment by jgriffiths** [comment]
> Issuance prevout has two parts: a 32-byte prevoutTxid (transaction hash) and a 32-bit little endian unsigned integer prevoutIndex (zero-based index of the transaction output). 

This doesn't explain what issuance prevout is, presumably its the issuance tx itself? Perhaps better to drop it entirely and document the 2 fields as something like:
```
prevoutTxid: 32 byte txid of the issuance transaction for the asset.
prevoutIndex: 32 bit little endian integer containing the output index of the asset issuance transaction.
```
(assuming I understand these fields correctly)

**Comment by jgriffiths** [comment]
> It is preceded by contractLen containing size of the contract in bytes as a compact size unsigned integer.

This is redundant (its in the `<valuedata>` column beside it).

**Comment by miketlk** [comment]
> > The fields added for PSET are only allowed when the PSBT version is 2.
> 
> Redundant, PSET only supports v2, and when v3+ are added this spec will remain valid if those versions are backwards compatible.

@jgriffiths thank you for reviewing ELIP. I will remove this statement.

**Comment by jgriffiths** [comment]
> The asset metadata including contract and issuance prevout. The parameter contract is a JSON document, canonicalized to have its keys sorted lexographically. 

A number of issues with this:

1. The content of the JSON isn't defined. Is it intended to follow the Blockstream asset registry? If so there should be a link to the format definition there.
2. There is no facility to version the JSON format
3. This adds a requirement for a JSON parser which is not viable for HWW IMO (Liquid is already very hard to fit on constrained devices)
4. JSON has no requirement to order its keys and imposing one here has several drawbacks, such as libraries not supporting it for output or more likely not validating it on input.
5. In fact the different behavior of JSON parsers in this respect has caused a number of CVEs in the past, as e.g. duplicate keys can be handled in different orders allowing attackers to pass multiple keys and pass different data to different layers due to them using different libraries.

I have further objections to using JSON here but need clarification before I can raise them.

**Comment by miketlk** [comment]
> > it is stored as 3 bytes starting with 0xfd. For example, identifier of PSBT_ELEMENTS_HWW_GLOBAL_ASSET_METADATA is a sequence of 9 bytes 0xfc0470736574fd0100.
> 
> This sequence of bytes does not start with 0xfd.

The meaning was not the whole key starts with 0xfd, but only a 16-bit subkeytype part of it: fd0100 (last 3 bytes of this example). But I do agree that this statement is quite vague. Anyway, considering your proposal on using a dedicated prefix, it will be removed from the spec.

**Comment by jgriffiths** [comment]
> The meaning was not the whole key starts with 0xfd, but only a 16-bit subkeytype part of it: fd0100 (last 3 bytes of this example). But I do agree that this statement is quite vague. Anyway, considering your proposal on using a dedicated prefix, it will be removed from the spec.

Sorry, I'm short on time so I'm being quite terse, my apologies.

**Comment by jgriffiths** [comment]
> Updater

The updater should also remove data for assets that become unreferenced by a transaction if it removes inputs from the PSET.

**Comment by miketlk** [comment]
> > Subtype of the fields belonging to the hardware wallet extensions always start with 0x0100 base to avoid potential collisions when new fields will be added to the mainstream PSET standard
> 
> PSBT/PSET already has a mechanism to avoid collisions, using the proprietary field prefix. This spec should use another field prefix than "pset" and number its fields from zero.

I do agree it would be better to use a separate prefix as a better way to avoid possible collisions. How about "psethww"? Do you think it is safe to assume that none of the existing implementations will consider records with unknown prefix as "alien" and remove them from PSET, rather than just ignoring them and passing through?

**Comment by jgriffiths** [comment]
> But if any of the asset metadata fields is present, the Signer must verify it by re-computing the asset tag from the given contract and prevout

Should probably link to the documentation describing how to do this here.

**Comment by jgriffiths** [comment]
> Test Vectors

A test vector of a faked/invalid contract would be invaluable for implementations to verify that their validation is correct.

**Comment by jgriffiths** [comment]
> Backwards Compatibility

This section is only valid if a different extension key is used. Otherwise, elements versions prior to about v22 will mishandle (or crash, I don't recall) PSETs using this ELIP.

**Comment by jgriffiths** [comment]
> I do agree it would be better to use a separate prefix as a better way to avoid possible collisions. How about "psethww"?

Works for me although I probably prefer "pset_hww".

>Do you think it is safe to assume that none of the existing implementations will consider records with unknown prefix as "alien" and remove them from PSET, rather than just ignoring them and passing through?

I only have experience with 3 implementations: elements, rust-elements and wally. Of the 3 all handle proprietary data correctly. Given that e.g. c-lightning uses this mechanism to attach processing data to lightning payments, and that the lightning ecosystem involves a bunch of different implementations, I'd say its safe, yes.

Although of course, losing this information in this case is not fatal in any case according to the ELIP itself :)

**Comment by jgriffiths** [comment]
> But if any of the asset metadata fields is present, the Signer must verify it by re-computing the asset tag from the given contract and prevout.

Probably worth stating that the signer must validate all of the metadata present, even if it is not for an input that the user is signing for.

In order to reduce resource exhaustion attacks, I would probably require that an ELIP100 compatible processor must not include asset metadata for assets that are not present in the PSET. (this is separate from concerns about the json contract format which we can discuss once its more clear).

**Comment by jgriffiths** [comment]
Overall, this appears to give the same level of assurances about the metadata that a HWW will display as the asset registry does. Last time I looked at the asset registry I had some concerns that it did not protect sufficiently against phishing etc. If you can update per the above comments to link to the current json format and verification procedure I can review those too, thanks.

**Comment by miketlk** [comment]
> > `<compact size uint contractLen><contract><32-byte prevoutTxid><32-bit little endian uint prevoutIndex>`
> 
> As a general rule, prefer to have fixed size data before variable length data, it makes decoding easier since otherwise all fields need extra math to adjust their offset at runtime.

Do you think it's a critical design flaw? Looking at other Liquid and Bitcoin specs it looks like this rule is not always followed, in some places there is a mix of fixed and variable length fields. This modification is more significant than others, as the parser state machine would need to be re-composed.

The idea was to optimize memory use on the resource-restricted platforms by choosing the order of fields as they are used in processing. After hashing the contract, the SHA context may be reused to compute a double hash of prevout. Currently, my implementation on Ledger Nano S processes asset metadata on the fly as it is streamed from the host app. There is no enough memory to store this record entirely anyway, to benefit from having fixed offsets.

**Comment by miketlk** [comment]
> > Issuance prevout has two parts: a 32-byte prevoutTxid (transaction hash) and a 32-bit little endian unsigned integer prevoutIndex (zero-based index of the transaction output).
> 
> This doesn't explain what issuance prevout is, presumably its the issuance tx itself? Perhaps better to drop it entirely and document the 2 fields as something like: ...

Ok, I will update the description to make it more clear.

**Comment by jgriffiths** [comment]
> Do you think it's a critical design flaw?

Not at all, just a general comment that I prefer simpler parsing if possible. I should have prefixed this comment with `nit:`, its not a deal breaker and if you have justification for it then its reasonable as-is I think.

**Comment by miketlk** [comment]
> > It is preceded by contractLen containing size of the contract in bytes as a compact size unsigned integer.
> 
> This is redundant (its in the `<valuedata>` column beside it).

The intent was to describe (or at least to mention) all the fields included in `<valuedata>` to make the spec more formal. Even if their meaning is obvious from the context. If the Elements documentation guidelines recommend preferring the brevity, I can remove it.

**Comment by jgriffiths** [comment]
> Even if their meaning is obvious from the context. If the Elements documentation guidelines recommend preferring the brevity, I can remove it.

Its more that it reads differently to the PSBT spec, although I recognise that there isn't enough consistency there already. If you worded it as e.g. `PSBT_IN_TAP_BIP32_DERIVATION` is in https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki it would read better (something like `A compact size unsigned integer representing the size of the JSON document followed by the JSON document`).

**Comment by miketlk** [comment]
> > The asset metadata including contract and issuance prevout. The parameter contract is a JSON document, canonicalized to have its keys sorted lexographically.
> 
> A number of issues with this:
> 
> 1. The content of the JSON isn't defined. Is it intended to follow the Blockstream asset registry? If so there should be a link to the format definition there.

Yes, the JSON contract should be exactly as stored in Blockstream registry because it is hashed to compute asset entropy and then asset tag. I don't know if its detailed specification exists somewhere. It was briefly described on Liquid Developer Documentation portal, but after a recent update it's gone. Current implementation on Nano S expects two level-1 properties: "ticker" and "precision".

> 2. There is no facility to version the JSON format

That's correct. But for the purposes of the HWW app it's not critical, as long as the contract contains ticker and precision and its hash results in the correct asset tag. 

> 3. This adds a requirement for a JSON parser which is not viable for HWW IMO (Liquid is already very hard to fit on constrained devices)

In the current Nano S app, I implemented a very minimalistic 1-dimentional JSON parser, ignoring nested objects and arrays. It turned out to be pretty simple and not requiring to hold entire JSON in memory, processing it as a stream and extracting only the needed values. Contract hash is also computed on the fly.

> 4. JSON has no requirement to order its keys and imposing one here has several drawbacks, such as libraries not supporting it for output or more likely not validating it on input.

In the ELIP 100, there is an explicit requirement that the JSON must be canonicalized with lexicographical order of keys. This is a strict form of JSON defined in RFC 8785 which theoretically forces all implementation to produce bit-exact results. In the Blockstream asset registry, contracts are stored already in the right form, so the host software is not obligated to do any normalization.

> 5. In fact the different behavior of JSON parsers in this respect has caused a number of CVEs in the past, as e.g. duplicate keys can be handled in different orders allowing attackers to pass multiple keys and pass different data to different layers due to them using different libraries.

Current implementation on Nano S checks for duplication of keys for the properties it handles. A duplicating key results in parsing error and the HWW app refuses to sign such PSET. Also, if the JSON contract has been altered in any way by the attacker, the resulting asset tag will be wrong because it's cryptographically bound to the hash of the contract.

> I have further objections to using JSON here but need clarification before I can raise them.

Of course, the JSON format has its drawbacks and limitations, but it seems there aren't any alternatives in the scope of HWW app because. It needs the original contract whose hash was used to issue the asset. If the contract data is presented in some other form along with a pre-computed contract hash, the HWW app would not be able to verify this data cryptographically.

**Comment by miketlk** [comment]
> > But if any of the asset metadata fields is present, the Signer must verify it by re-computing the asset tag from the given contract and prevout
> 
> Should probably link to the documentation describing how to do this here.

Unfortunately, I didn't find a document describing this process. I used source code of Elements core and Rust Elements as a reference. I can describe asset computation in ELIP 100, but I'm not sure that it's the right scope for it. I think it would be better to define general things in basic docs, and then reference them in application-specific docs like this ELIP.

**Comment by miketlk** [comment]
> > Test Vectors
> 
> A test vector of a faked/invalid contract would be invaluable for implementations to verify that their validation is correct.

That's a good idea to add test vectors. Not even for the contract, but probably for the whole metadata record as well. By the way, the main criteria for the wrong contract is a mismatch between computed asset tag and the asset tag included in PSET key.

**Comment by miketlk** [comment]
> Overall, this appears to give the same level of assurances about the metadata that a HWW will display as the asset registry does. Last time I looked at the asset registry I had some concerns that it did not protect sufficiently against phishing etc. If you can update per the above comments to link to the current json format and verification procedure I can review those too, thanks.

As I mentioned in earlier replies, the HWW app can verify cryptographically the asset metadata by using it to compute the asset tag. This is the same process as Elements core uses when issuing the asset. If the resulting asset tag matches the one contained in the transaction (and in the metadata key), the HWW app considers the metadata as valid.

**Comment by jgriffiths** [comment]
> statemends

nit: statements

**Comment by jgriffiths** [comment]
> there aren't any alternatives in the scope of HWW app because it needs the original contract whose hash was used to issue the asset

I've had time to think about this a little more.

First, it seems that JSON can not be eliminated since that ship has sailed. Given that, discussions about the JSON format/what should be allowed in it, and documentation of the verification procedure should be punted to the registry and linked here. This spec is not the place to re-state or re-litigate those decisions, but it is also not an implementable spec without a link describing the format and how verification is done. As such I can't ack it if the underlying data it proports to allow verification of is not defined, even if that is outside of the scope of this ELIP.

I think the registry JSON format should consider the needs of HWW in verifying the asset metadata including changes such as limiting the allowable JSON elements to flat string key,value pairs only, and disallowing all JSON processing characters (`"{}:\\,\n`) inside strings. 

I have grave concerns about the utility and safety of the asset registry data since my original comments on verifying for duplicate data/phishing a year ago were largely ignored. If it is true that the documentation is lacking or non-existent then this is simply unacceptable due to the risk of naive users not understanding the level of trust that can be placed in the asset metadata.

**Comment by jgriffiths** [comment]
Secondly, on second thought I realize it is not true that providing this metadata inside the PSET provides a similar level of assurance to fetching this data from the registry out of band.

The behavior as I understand it simply allows the HWW to verify that the asset tag is in fact a commitment to the asset metadata and issuance tx.

This does not provide any further assurance that the data itself is valid or trustworthy, which is a concern if the intention is for the HWW to display it to the user for signing confirmation.

In Green, we fetch the asset metadata from the registry OOB and pass it to Jade. Provided the user is not running a compromised Green app, this allows us to verify that the metadata is present in the registry and thus has been vetted somewhat, notwithstanding my concerns about the nature and laxity of this verification. At least some checks such as validating that the issuing domain is owned by the asset issuer can be repeated in the client app. Other checks like e.g. ensuring the ticker is not a duplicate/visual duplicate are possible even if not implemented yet.

Passing this data in the PSET in comparison does not ensure that the minimal registry checks have been performed, or even that the metadata exists in the registry. What is stopping an attacker from issuing an asset using e.g. Tethers current metadata, and simply not publishing it to the registry (where it should fail validation)? The asset id would be incorrect, but the metadata would look the same to a naive user signing a tx, who is more likely to check the displayed ticker and skip checking the asset id.

Witholding the metadata from the registry also allows for targeting the precision (and thus amount) that a signer displays, even if the asset isn't attempting to pass itself off as another asset.

**Comment by jgriffiths** [comment]
If the documentation of the JSON format, its limitations and what can be inferred from it under given assumptions is punted elsewhere, then this spec becomes a lot simpler and many of my comments can move to that documentation instead of here..

In the Green case, we are probably not going to stop fetching the metadata from the registry directly, and if/when we support this ELIP we will probably just verify that any contract data present matches the registry contents. I'm agnostic on whether Jade should continue to receive this data OOB or get it from the PSET, that's most likely an implementation decision for @jdriver.

**Comment by jgriffiths** [comment]
A possible improvement on the above is for the registry to provide a signature for the asset tag, and embed the registry pubkey into the HWW firmware. Ignoring key compromise and any registry-eviction concerns, this would allow the HWW to validate that the metadata had been registered and (with some trust assumptions) ensure that the further validation imposed by the registry had been performed, which IMO is an improvement.

**Comment by miketlk** [comment]
> I think the registry JSON format should consider the needs of HWW in verifying the asset metadata including changes such as limiting the allowable JSON elements to flat string key,value pairs only, and disallowing all JSON processing characters (`"{}:\\,\n`) inside strings.

Sure, this could ease things for the HWW app developers, of course. But thousands of assets are already issued, and their contracts cannot be modified anymore, because of being bound to the asset tag. So, we need to deal with what we have. Nano S app handles escape symbols correctly and skips everything except the key-value pairs. Thus, it's manageable, although requiring more tests to cover all the cases.

**Comment by miketlk** [comment]
> Secondly, on second thought I realize it is not true that providing this metadata inside the PSET provides a similar level of assurance to fetching this data from the registry out of band.
> This does not provide any further assurance that the data itself is valid or trustworthy, which is a concern if the intention is for the HWW to display it to the user for signing confirmation.

As far as I know, the security model of any cold wallet tolerates signing an invalid transaction because it's not possible to validate it fully without having a complete blockchain. Cold wallet simply cannot detect, for example, a double-spend transaction or a transaction belonging to the wrong chain. Usually, it's not seen as an issue because invalid transactions have no chance to be accepted by the network (validating nodes for Liquid). If an attacker replaces an asset tag with another one they generated from a fake contract, this transaction will not be accepted by validators, simply because the victim has no UTXO with such asset on the blockchain (nothing to spend). Also, probably the validators will check the Blockstream registry as well to reject "unofficial" assets (but I'm not sure about this policy). 

> In Green, we fetch the asset metadata from the registry OOB and pass it to Jade. Provided the user is not running a compromised Green app, this allows us to verify that the metadata is present in the registry and thus has been vetted somewhat, notwithstanding my concerns about the nature and laxity of this verification. At least some checks such as validating that the issuing domain is owned by the asset issuer can be repeated in the client app. Other checks like e.g. ensuring the ticker is not a duplicate/visual duplicate are possible even if not implemented yet.

The same approach can be used with other HWWs as well, expecting ELIP 100 metadata inside PSET. The Green app can fetch from registry and add the asset metadata just before passing PSET to the HWW app for signing. This "enriched" PSET will not be transmitted to other multisig parties to not encourage them to trust this information. If the Green app receives a PSET with asset metadata already built-in, it can verify this data using Blockstream registry or simply remove it if in doubt.
 
> Passing this data in the PSET in comparison does not ensure that the minimal registry checks have been performed, or even that the metadata exists in the registry. What is stopping an attacker from issuing an asset using e.g. Tethers current metadata, and simply not publishing it to the registry (where it should fail validation)? The asset id would be incorrect, but the metadata would look the same to a naive user signing a tx, who is more likely to check the displayed ticker and skip checking the asset id.

This will result in an invalid transaction, thus having a valid signature. Such a transaction is not going to be accepted by the validators, so an attacker will not be able to take much advantage of it. There is a higher risk if the victim gets paid in these fake assets, but in this process the HWW plays a limited role, just displaying a receive address.

**Comment by miketlk** [comment]
> A possible improvement on the above is for the registry to provide a signature for the asset tag, and embed the registry pubkey into the HWW firmware. 

It's technically possible, but in my opinion it would be better to get a confirmation first that this feature will be implemented in asset registry at some point. It seems to me that the current security model assumes that it's enough to have a one-way collision-resistant cryptographic relationship between an asset's metadata and its tag.

**Comment by miketlk** [comment]
Here's the PR #6, where the majority of the recommendations have been applied. I will add test vectors later when I update the Python code generating test data for the Nano S app.

**Comment by jgriffiths** [comment]
> It's technically possible, but in my opinion it would be better to get a confirmation first that this feature will be implemented in asset registry at some point.

Yes, I think punting the description, rationale and security assumptions for the registry data outside of the scope of this ELIP is needed, since this is the wrong place to discuss them.

> This will result in an invalid transaction, thus having a[n in]valid signature. 

This is only true if the actual asset id in the tx is the real tether asset id. If its a bogus asset I made up myself using the tether ticker and domain (and then didn't publish to the registry), it will result in a valid tx for that asset, the HWW will just display the ticker as tether which could easily be enough to convince naive users to swap valuable assets for it.

The registry at least nominally protects users from this by its own validation (at least it should, last I asked there was no plan to prevent duplicate tickers). But to prove that such validation has taken place, you need to prove that the JSON was registered, i.e. you need a signature from the registry.

**Comment by jgriffiths** [comment]
> Nano S app handles escape symbols correctly and skips everything except the key-value pairs. Thus, it's manageable, although requiring more tests to cover all the cases.

Given the limitations of rolling out firmware updates to fix bugs, and the difficulty in correctly parsing arbitrary JSON, I'd much rather see a model where the JSON is treated simply as a preimage and no parsing is attempted on device, and a fixed structure of `<json hash><asset tag><precision><domain><registry sig>` is included to pull this data from. All of the information that will be presented to the user is only trustworthy if we trust that a) the registry has validated it and b) the registry does reasonable validation to prevent malicious metadata.

**Comment by miketlk** [comment]
> If its a bogus asset I made up myself using the tether ticker and domain (and then didn't publish to the registry), it will result in a valid tx for that asset, the HWW will just display the ticker as tether which could easily be enough to convince naive users to swap valuable assets for it.

Yes, within the HWW realm, this transaction will be considered valid if an attacker will also create a fake UTXO for this nonexistent asset and insert it into PSET. This is true because the typical HWW doesn't look behind in history earlier than 1 transaction. But for a validating node, this transaction is not valid because there is no such UTXO in the blockchain, and thus it will be rejected.

> > simply because the victim has no UTXO
> 
> The victim is not expected to have such a UTXO in this attack, the idea is to convince them to swap a valid asset for a bogus one masquerading as a legitimate asset (e.g. swapping L-BTC for tether in this example).

It seems I don't understand the point of this attack after all. Do you mean some off-chain operation which bypasses the consensus mechanism of validators and for which a non-valid but correctly signed transaction would have some use?
 
> The registry at least nominally protects users from this by its own validation (at least it should, last I asked there was no plan to prevent duplicate tickers). But to prove that such validation has taken place, you need to prove that the JSON was registered, i.e. you need a signature from the registry.

It seems to me that this centralized approach doesn't fit well with the original idea of the Liquid Network with independent validators. Making Blockstream asset registry a single source of truth creates a single point of failure and rises the infrastructural risks.

**Comment by miketlk** [comment]
> Given the limitations of rolling out firmware updates to fix bugs, and the difficulty in correctly parsing arbitrary JSON, I'd much rather see a model where the JSON is treated simply as a preimage and no parsing is attempted on device, and a fixed structure of `<json hash><asset tag><precision><domain><registry sig>` is included to pull this data from. All of the information that will be presented to the user is only trustworthy if we trust that a) the registry has validated it and b) the registry does reasonable validation to prevent malicious metadata.

This approach would work more effectively on HWW, no doubt. But it seems to me that it doesn't have a great chance of adoption because it would require creating a separate registry of asserts specifically for HWW. Also, from my HWW developer's perspective, it will be less secure because you can no longer verify whether the metadata matches the asset label, and have to trust the signature of some authority.

**Comment by jgriffiths** [comment]
> This is true because the typical HWW doesn't look behind in history earlier than 1 transaction. But for a validating node, this transaction is not valid because there is no such UTXO in the blockchain, and thus it will be rejected.

We seem to be talking past each other here. AFAICS, I can issue an asset using someone elses asset json, and nothing in Liquid will stop me. I then have a valid UTXO on the chain and asset json that matches an existing asset. The only difference between the real asset and mine is that the actual asset tag is different, since it includes the issuance output in computing it. There is no node-level validation that will prevent this (the contract details are not validated by nodes beyond that the contract json hashed with the outpoint equals the asset id).

> It seems I don't understand the point of this attack after all. Do you mean some off-chain operation which bypasses the consensus mechanism of validators and for which a non-valid but correctly signed transaction would have some use?

Nothing like that. This is a standard, valid on-chain swap transaction between my asset (with identical metadata to an existing asset, except for the asset tag). If signed, it will pass consensus validation and be mined.

The attack is that without online validation, e.g. that the domain in the metadata contains the asset, and/or that the registry has seen it, displaying the ticker and domain to the user can trick them into believing that my fake asset is the real one.

**Comment by miketlk** [comment]
> This is a standard, valid on-chain swap transaction between my asset (with identical metadata to an existing asset, except for the asset tag). If signed, it will pass consensus validation and be mined.
> 
> The attack is that without online validation, e.g. that the domain in the metadata contains the asset, and/or that the registry has seen it, displaying the ticker and domain to the user can trick them into believing that my fake asset is the real one.

So, as I understand, this attack is specific to swap transaction only? In means that the HWW wallet when displaying somebody else's inputs may indicate a bogus asset impersonating some original one. I cannot right now suggest a solution that could be implemented solely within the capabilities of HWW. The only thing that could be helpful is a list of most used assets hard-coded in HWW firmware. It is already implemented in Nano S app, but this: (1) will not cover all cases, (2) requires an advanced algorithm looking through the asset list for the visually or semantically similar tickers (like USDt, USDT, US$t, TETHER, UStether, etc...). Or, the users should ensure their host software (desktop, mobile) is genuine and unmodified before deciding to make a swap transaction.

**Comment by miketlk** [comment]
Added test vectors to PR #6

**Comment by jgriffiths** [comment]
>So, as I understand, this attack is specific to swap transaction only?

This affects any multi-party tx, swaps are just the simplest case. Things like joint escrow are similarly vulnerable (where the other participants are escrowing bogus assets instead of the agreed on one(s)). However you can imagine a poorly implemented/manual system verifying proof of funds or deposit that is coded to check the ticker instead of the asset id also being affected.

The key idea here is not to present a false sense of validity to bogus assets by displaying their ticker as though that represents the actual ticker when in fact there is no such assurance. This is why being able to link to a document that describes what validation is done and critically what this means in terms of the data that is validated is critical.

> means that the HWW wallet when displaying somebody else's inputs may indicate a bogus asset impersonating some original one.

Yes, exactly.

> I cannot right now suggest a solution that could be implemented solely within the capabilities of HWW. The only thing that could be helpful is a list of most used assets hard-coded in HWW firmware. It is already implemented in Nano S app, but this: (1) will not cover all cases

This is why I suggest allowing for external signatures. Rather than attempting to bake every asset type into the firmware (impractical for space reasons, and requires constant updates), simply bake in one or more pubkeys allowing to validate that the firmware maker has validated the data (to some extent which may differ depending on the metadata signer).

>, (2) requires an advanced algorithm looking through the asset list for the visually or semantically similar tickers (like USDt, USDT, US$t, TETHER, UStether, etc...).

Or manual inspection, or a registry that requires a deposit to register. etc. etc. Its a spectrum. The degree of checking that is done is a feature of whoever signs the metadata, and the HWW should indicate this to the user when signing. Displaying the real asset and my bogus one identically when signing is asking for trouble. As far as this ELIP is concerned there should at least be some mechanism to indicate that some optional external validation has been performed IMO.

> Or, the users should ensure their host software (desktop, mobile) is genuine and unmodified before deciding to make a swap transaction.

This does not help. e.g. on a decentralised swap exchange I can offer my bogus asset for sale to trick naive users - if they accept the swap then even with full integrity of their SW/HWW chain I will have effectively stolen their funds.

**Comment by miketlk** [comment]
> The key idea here is not to present a false sense of validity to bogus assets by displaying their ticker as though that represents the actual ticker when in fact there is no such assurance. This is why being able to link this ELIP to a document that describes what validation is done and what this means in terms of the data that is validated is critical.

Probably it makes sense to display full asset tag in HEX if the metadata was obtained from PSET and not the hardcoded list.

> This is why I suggest allowing for external signatures. Rather than attempting to bake every asset type into the firmware (impractical for space reasons, and requires constant updates), simply bake in one or more pubkeys allowing to validate that a registry/firmware maker has validated the data (to some extent which may differ depending on the metadata signer).

It's a significant improvement, but it also has drawbacks. First, the lack of backward compatibility: to make the signatures meaningful after they are approved, HWW must refuse to process any asset metadata without such signature. Ideally there should be no firmware supporting unsigned metadata at all, so that there is no concern about the unequal level of risk for users who updated and those who did not update the firmware. Secondly, and more significantly in my opinion, the leak of a private key from one of these asset repositories immediately makes this protection ineffective. Moreover, given the irregular nature of HWW updates it will be problematic to ensure there are no more devices using the public key for which the private key has been compromised. Probably a multisig scheme could help, but it will make processing PSET terribly slow on resource constrained platforms like Nano S (it's not so fast already).

> 
> This does not help. e.g. on a decentralised swap exchange I can offer my bogus asset for sale to trick naive users - if they accept the swap then even with full integrity of their SW/HWW chain I will have effectively stolen their funds.

I meant that the host software (when it's not hacked) must be responsible for making a reasonable number of online checks before injecting asset metadata in PSET intended for HWW. That was the original idea: asset metadata is a _temporary_ field whose lifetime is limited to a single communication session with HWW. There is no reason to distribute PSET with metadata built-in. However, it is not forbidden in this ELIP. But now I understand that it should be stated explicitly. By the way, PSET has several fields with similar lifetime: PSBT_ELEMENTS_IN_EXPLICIT_VALUE, PSBT_ELEMENTS_IN_VALUE_PROOF etc.

**Comment by jgriffiths** [comment]
Agree that there are drawbacks and tradeoffs with various schemes including signing the metadata.

Also agree that the asset hex must be shown if the HWW cannot otherwise prove provenance.

Strongly agree that It should be explicitly stated if the metadata is not intended to be transmitted. I was not aware that this was the intent.

>  PSET has several fields with similar lifetime

This is not the case for the 2 fields you listed AFAIU. They are intended to be transmitted under certain circumstances, in particular in swap protocols where you do not wish to share blinding factors with the other party.

**Comment by miketlk** [comment]
Added security-related statements and precautions in commit f005e50e3017874d2fb666116325cbe8591fe20a

**Comment by jgriffiths** [comment]
I've reviewed the PR, LGTM, thanks for your efforts here!

**Comment by miketlk** [comment]
@jgriffiths, thank you for your suggestions in PR #6. They are implemented in c3e3325bb5e994b6a89f1bf04f22343c1f37ae5e and 37efbf468820f6155f3f2507a4ce61afdd6ee1c2.

**Comment by jgriffiths** [comment]
@miketlk Thanks, appreciate your work to clarify this stuff. I've acked these changes.

IMO the only thing missing is a link to the relevant json/registry docs. If there is no such documentation please create an issue on the registry to provide them, referencing this ELIP.
